core/src/appleMain/kotlin/CBPeripheralCoreBluetoothPeripheral.kt

```diff
@@ -264,9 +274,14 @@ internal class CBPeripheralCoreBluetoothPeripheral(
             centralManager.discoverServices(cbPeripheral, servicesToDiscover)
         }
 
-        cbPeripheral.services?.forEach { cbService ->
+        cbPeripheral.services?.filterIsInstance<CBService>()?.forEach { cbService ->
             connection.execute<DidDiscoverCharacteristicsForService> {
-                centralManager.discoverCharacteristics(cbPeripheral, cbService as CBService)
+                centralManager.discoverCharacteristics(cbPeripheral, cbService)
+            }
+            cbService.characteristics?.forEach { cbCharacteristic ->
+                connection.execute<DidDiscoverDescriptorsForCharacteristic> {
+                    centralManager.discoverDescriptors(cbPeripheral, cbCharacteristic as CBCharacteristic)
```

Comment:
```suggestion
            // Cast should be safe since `CBService.characteristics` type is `[CBCharacteristic]?`,
            // according to: https://developer.apple.com/documentation/corebluetooth/cbservice/characteristics
            @Suppress("UNCHECKED_CAST")
            val discoveredCharacteristics = cbService.characteristics as List<CBCharacteristic>?
            discoveredCharacteristics?.forEach { cbCharacteristic ->
                connection.execute<DidDiscoverDescriptorsForCharacteristic> {
                    centralManager.discoverDescriptors(cbPeripheral, cbCharacteristic)
```

Reply:
I'll push an update to the PR with the approach you suggested